packages/server/src/api/rest/index.ts

```diff
@@ -809,6 +829,89 @@ class RequestHandler extends APIHandlerBase {
         };
     }
 
+    private async processUpsert(
+        prisma: DbClientContract,
+        type: string,
+        _query: Record<string, string | string[]> | undefined,
+        requestBody: unknown,
+        modelMeta: ModelMeta,
+        zodSchemas?: ZodSchemas
+    ) {
+        const typeInfo = this.typeMap[type];
+        if (!typeInfo) {
+            return this.makeUnsupportedModelError(type);
+        }
+
+        const { error, attributes, relationships } = this.processRequestBody(type, requestBody, zodSchemas, 'create');
+
+        if (error) {
+            return error;
+        }
+
+        const matchFields = (requestBody as any).meta.matchFields;
+        const uniqueFields = Object.values(modelMeta.models[type].uniqueConstraints || {}).map((uf) => uf.fields);
+
+        if (
+            !uniqueFields.some((uniqueCombination) => uniqueCombination.every((field) => matchFields.includes(field)))
+        ) {
+            return this.makeError('invalidPayload', 'Match fields must be unique fields', 400);
+        }
+
+        const upsertPayload: any = {
+            where: this.makeUpsertWhere(matchFields, attributes, typeInfo),
+            create: { ...attributes },
+            update: {
+                ...Object.fromEntries(Object.entries(attributes).filter((e) => !matchFields.includes(e[0]))),
+            },
+        };
+
+        if (relationships) {
+            for (const [key, data] of Object.entries<any>(relationships)) {
+                if (!data?.data) {
+                    return this.makeError('invalidRelationData');
+                }
+
+                const relationInfo = typeInfo.relationships[key];
+                if (!relationInfo) {
+                    return this.makeUnsupportedRelationshipError(type, key, 400);
+                }
+
+                if (relationInfo.isCollection) {
+                    upsertPayload.create[key] = {
+                        connect: enumerate(data.data).map((item: any) =>
+                            this.makeIdConnect(relationInfo.idFields, item.id)
+                        ),
+                    };
+                    upsertPayload.update[key] = {
+                        connect: enumerate(data.data).map((item: any) =>
```

Comment:
I'm wondering if the semantic for the "update" case should use "set" instead of "connect" here - replace the entire connection collection instead of adding new ones?